### DISCUSSION TITLE
Crop image based on content placement with rolling window

### DISCUSSION THREAD

**Discussion by cmd8** [ImageMagick/ImageMagick#5073] [Help] [answered]
Hey all! I have some experience working with imagemagick on a basic to intermediate level, not sure if the following is possible and if it is I'd appreciate if someone could point me toward the right direction.

Given, a number of image files that have some identical content in them in different places like on the image below. Source of the content in the blue squares is available.

![1](https://user-images.githubusercontent.com/13213650/164955152-2f737a5f-f5af-4464-924f-65ec41f96b9a.jpeg)

Is it possible for imagemagick to perform a rolling window content search and then crop the images so that the content of the blue squares would be placed in the middle of a new image like in the example below?

![2](https://user-images.githubusercontent.com/13213650/164955216-6b29f5d3-4cac-40ea-bf0a-f75790a446af.jpeg)

Any guidance would help!

**Comment by fmw42** [comment, marked as answer]
Please define "rolling window search"?  Do you have a template that you want to find in your images?  Is the background constant color (gray or some other constant color)?  What is your ImageMagick version and platform/OS?

**Reply by cmd8** [reply]
> Please define "rolling window search"? 

Rolling window or sliding window is just a terminology I used to mention one possible approach on finding content inside images e.g. https://pyimagesearch.com/2015/03/23/sliding-windows-for-object-detection-with-python-and-opencv/

> Do you have a template that you want to find in your images?

Yes, the template is available, as I mentioned in the original post: "Source of the content in the blue squares is available"

> Is the background constant color (gray or some other constant color)? 

Can be anything really, any background

> What is your ImageMagick version and platform/OS?

Version: ImageMagick 7.1.0-29 Q16-HDRI x86_64
Platform: MacOS 12.3.1

**Reply by fmw42** [reply]
See the compare tool with -subimage-search option. The compare tool is a sliding window template matching tool, so it can be slow if you have a large image. 

https://imagemagick.org/script/compare.php
https://legacy.imagemagick.org/Usage/compare/

example:

`magick compare -metric RMSE -subimage-search logo.png wizard.jpg similarity.tif`

where logo.png is the larger image and wizard.jpg is the template image. I would recommend that if the image is large, then use -metric NCC rather than RMSE as it is much faster.

Also see the old post at https://legacy.imagemagick.org/discourse-server/viewtopic.php?f=1&t=14613&p=51076&hilit=compare+metric#p51076, but add -subimage-search (since at the time of the old post that was not needed)

**Reply by cmd8** [reply]
Though slow it'd work, thanks!

**Reply by fmw42** [reply]
If on IM 7, use -metric NCC. It is much faster.
